### ISSUE TITLE
Reconsider license to allow use by free software projects

### ISSUE THREAD

**Issue by billziss-gh** [billziss-gh/winfsp#32]
WinFsp is currently GPLv3:

https://github.com/billziss-gh/winfsp/blob/v1.0RC1/License.txt

The license has this header:

    The WinFsp project is Copyright (C) Bill Zissimopoulos. It is licensed
    under the terms of the GPLv3. The full text of this license follows
    below. Commercial licensing options are also available: Please contact
    Bill Zissimopoulos <billziss at navimatics.com>.

In order to ease adoption by open-source projects that are not GPLv3, it might be worthwhile to explicitly allow such projects to use and distribute WinFsp without having to comply with the GPLv3 themselves. Closed-source projects would still require a commercial license.

Proposed new header:

    The WinFsp project is Copyright (C) Bill Zissimopoulos. It is licensed
    under the terms of the GPLv3.

    In addition to the terms of the GPLv3, it is explicitly allowed that free
    software projects (as per the FSF definition) may use unmodified binary
    copies of the software (as released by the WinFsp project) provided that
    they include a WinFsp copyright notice in the project's documentation
    and/or "About" box, etc.

    The full list of free software licenses can be found at the URL below.
    https://www.gnu.org/licenses/license-list.html

    The licenses that qualify for this exception are listed under "GPL-
    Compatible Free Software Licenses" and "GPL-Incompatible Free
    Software Licenses.

    Unless you have a commercial license you may not: (1) use WinFsp from
    a non-free software project/component, and (2) distribute WinFsp with a
    non-free software project/component. Please contact Bill Zissimopoulos
    <billziss at navimatics.com> for commercial licensing.

**Comment by okuoku** [comment]
(Strictly speaking: in FSF-sense, "free software" does *not* immediately mean "software that licensed/distributed under free software license"; eg) one can distribute BSD licensed (modified) software without any sources -- software without source code does not provide freedom 1 of The Free Software Definition.)

I'm not sure why Cryptomator folks considered their works are not compatible with current GPLv3.. Their license term is X11 which is compatible with GPL -- they can distribute their client binary with winfsp without changing their source license as long as the binary distribution contain pointer to winfsp sources. (Perhaps they don't have noticed the license changed from AGPL?)

Additional terms relative to the GPL is almost always stated as an "exception". e.g. MariaDB(MySQL)'s FLOSS exception https://mariadb.com/kb/en/mariadb/mariadb-license/#the-floss-exception XSD has slightly modified variant which cites FSF license list http://www.codesynthesis.com/projects/xsd/FLOSSE .

**Comment by billziss-gh** [comment]
@okuoku thank you very much for your comments and your very interesting pointers.

> I'm not sure why Cryptomator folks considered their works are not compatible with current GPLv3.

I think the issue is that if you include GPL code into an MIT licensed project, you are still required to publish the whole under the GPL. [At least that is my understanding.]

> Additional terms relative to the GPL is almost always stated as an "exception".

These are very useful pointers. Thank you!

**Comment by okuoku** [comment]
> you are still required to publish the whole under the GPL

It's only applicable to the binary distribution. But, yes, binary distributions will be governed under GPL so they are no longer be able to link against non-GPL-compatible libraries such as OpenSSL because of winfsp.. I agree it is a bit too restrictive for large software projects -- but it can be explicitly allowed by adding some linking exceptions.

Perhaps Cygwin's previous licensing term is the closest thing what you want:
https://web.archive.org/web/20151223215900/http://cygwin.com/licensing.html
> As a special exception to GPLv3+, Red Hat grants you permission to link software whose sources are distributed under a license that satisfies the Open Source Definition with libcygwin.a, without libcygwin.a itself causing the resulting program to be covered by GPLv3+.
it requires source distribution and licensed under open source license.

Recently Red Hat terminated commercial licensing of Cygwin so they re-licensed it as LGPL + linking exception ( https://cygwin.com/licensing.html ).

**Comment by overheadhunter** [comment]
So you want something like the LPGL but restricting use in projects with non-published source code, did I get that right?

**Comment by dumblob** [comment]
I agree with the comments of @okuoku . Additionally I have a few quick comments on the licensing header itself.

~~~~
In addition to the terms of the GPLv3, it is explicitly allowed that free
software projects (as per the FSF definition) may use unmodified binary
copies of the software (as released by the WinFsp project) provided that
they include a WinFsp copyright notice in the project's documentation
and/or "About" box, etc.
~~~~
* It's not specified what is the copyright notice, how it should look like, what should be included. Look e.g. at BSD - there it's crystal clear what is the copyright notice.
* `provided that they include a WinFsp copyright notice in the project's documentation and/or "About" box, etc.` - it's not clear whether you're requiring the notice in the project's documentation (e.g. for project contributors) or in the end-user-facing documentation if any (the `if any` is also important as the project might have no user-facing documentation)
* `etc.` means absolutely nothing at court
* The `and/or` construct makes it hard to judge which one to choose in difficult situations and under weird conditions - I would avoid it or specify separately when each of the options should be used.
* The `"About" box` construct is misleading and accounts only for a small subset of possibilities where to put the notice. Imagine a project without any "About" box. Then there are basically two possibilities:
    1. Either the project will ignore it (because of the `and/or` ambiguity by choosing just `or`).
    1. Or the project will need to introduce some "About" box just because of WinFsp. Now the authors will ask how did you actually imagine an "About" box? And how not? Comply to this requirement is not easy in modern (e.g. mobile-friendly) UIs. Not speaking about universal UIs of multiplatform projects which will probably be the majority as WinFsp offers full FUSE compatibility.

    I would rephrase it to something like "...provided that they include the copyright notice `WinFsp, copyright (c) 2017 Bill Zissimopoulos, all rights reserved.` in the user interface of the project and in case there is any user-facing documentation, include the copyright notice also there. It's forbidden to deliberately make the copyright notice hardly reachable by the user or hidden." (even though you should give some thought to the word `deliberately` - it's a double-edged sword).

Also, based on the current licensing practices, usually a user-facing copyright notice is required by the author. Therefore if the project doesn't have any user-facing documentation and chooses the `or` from the `and/or` phrase, it doesn't need to include this notice anywhere or just in the contributors' documentation (which is usally read by just programmers).

~~~~
The full list of free software licenses can be found at the URL below.
https://www.gnu.org/licenses/license-list.html

The licenses that qualify for this exception are listed under "GPL-
Compatible Free Software Licenses" and "GPL-Incompatible Free
Software Licenses.

Unless you have a commercial license you may not: (1) use WinFsp from
a non-free software project/component, and (2) distribute WinFsp with a
non-free software project/component. Please contact Bill Zissimopoulos
<billziss at navimatics.com> for commercial licensing.
~~~~

The first two paragraphs are highly restrictive. No company will check the list of licenses every single time there should be any change. Also this disallows companies using their own free licenses (which are still compliant with Free Software). Also consider compliance not with Free Software as GNU defines it (because it's overly restrictive copyleft), but rather the OSI definition of [Open Source](https://opensource.org/osd ) (which would require changing `free software` and `non-free software` phrases to `Open Source software` and `non- Open Source software` or equivalents).

These two paragraphs in conjuction with the third one will make WinFsp unusable for many companies. I would rephrase it to something like:

~~~~
The full list of free software licenses can be found at the URL below.

https://www.gnu.org/licenses/license-list.html

The licenses that qualify for this exception are those listed in
"GPL-Compatible Free Software Licenses" and "GPL-Incompatible Free
Software Licenses" sections and those explicitly approved by the
contact person for WinFsp matters (see below).

Unless you have a commercial license you may not: (1) use WinFsp from
a non-free software project/component, and (2) distribute WinFsp with
a non-free software project/component.

Please contact Bill Zissimopoulos <billziss at navimatics.com>: (1) for
commercial licensing, (2) if any URL is not reachable or (3) if you would
like to request any licensing exception.
~~~~

**Comment by billziss-gh** [comment]
Thank you everyone for taking the time to comment. I apologize for my late response.

@okuoku wrote:
> Perhaps Cygwin's previous licensing term is the closest thing what you want:
https://web.archive.org/web/20151223215900/http://cygwin.com/licensing.html

I agree, this actually looks very good.

About the only thing that I dislike is that it does not allow redistribution of the Cygwin DLL. In the WinFsp case I would like to allow redistribution of unmodified releases (as made by the WinFsp project). This would allow a project that makes its source available under one of the approved licenses to ship a release binary/installer that includes the WinFsp installer.

@overheadhunter wrote:
> So you want something like the LPGL but restricting use in projects with non-published source code, did I get that right?

Yes on the second part. I would still want to keep the license GPL rather than LGPL.

> Edit: What about commercial licensing for proprietary projects? I am asking because we're offering some dual-licensed libs, too (for folks who want to create proprietary projects while using our GPL libraries).

Yes, I do intend to provide commercial licensing for proprietary projects. This is not because I intend to convert WinFsp into a business (there are very few file system developers in this world :-) ). It is simply because WinFsp has taken a considerable amount of time and energy to create.

@dumblob wrote:
> Additionally I have a few quick comments on the licensing header itself.

I agree with most of your comments regarding your licensing header. I will attempt to clarify based on your feedback.

> The first two paragraphs are highly restrictive. No company will check the list of licenses every single time there should be any change. Also this disallows companies using their own free licenses (which are still compliant with Free Software). Also consider compliance not with Free Software as GNU defines it (because it's overly restrictive copyleft), but rather the OSI definition of Open Source (which would require changing free software and non-free software phrases to Open Source software and non- Open Source software or equivalents).

I explicitly chose the FSF definition / list simply because I am more familiar with it. I will have to study the OSI definition and understand the differences for me to make this change.

Perhaps it is best to include the actual list of allowed licenses in the header itself to avoid confusion re: which licenses are allowed or not. The MariaDB license exception seems to be doing this.

> These two paragraphs in conjuction with the third one will make WinFsp unusable for many companies. I would rephrase it to something like:
>
> [snip]
>
>     Please contact Bill Zissimopoulos <billziss at navimatics.com>: (1) for
>     commercial licensing, (2) if any URL is not reachable or (3) if you would
>     like to request any licensing exception.

I probably would not want (2) or (3), because I would rather not receive a lot of lawyer email asking me to review their special version of an open source license. IANAL and not qualified to do such reviews (not to mention I do not have the time or inclination to receive such mails).

**Comment by billziss-gh** [comment]
PS: In case I did not make it clear I intend to rewrite the license header/exception based on your feedback/links.

**Comment by billziss-gh** [comment]
I have below an updated version of the license header/exception:

    The WinFsp project is Copyright (C) Bill Zissimopoulos. It is licensed
    under the terms of the GPLv3.

    As a special exception to GPLv3, Bill Zissimopoulos grants additional
    permissions to Free/Libre and Open Source Software ("FLOSS") without requiring
    that such software is covered by the GPLv3.

        1. Permission to link with a platform specific version of the WinFsp DLL
        (currently winfsp-x64.dll or winfsp-x86.dll).

        2. Permission to distribute unmodified binary releases of the WinFsp
        installer (as released by the WinFsp project).

        These permissions (and no other) are granted provided that the software:

        1. Is distributed under a license that satisfies the Free Software
        Definition Version 1.141 (https://www.gnu.org/philosophy/free-sw.en.html)
        or the Open Source Definition Version 1.9 (https://opensource.org/osd).

        2. Includes the copyright notice "WinFsp - Windows File System Proxy,
        Copyright (C) Bill Zissimopoulos" and a link to the WinFsp repository in
        its user-interface and any user-facing documentation.
    
        3. Is not linked or distributed with proprietary (non-FLOSS) software.
        [You cannot mix FLOSS and proprietary software while using WinFsp under
        this special exception.]

    Commercial licensing options are also available: Please contact
    Bill Zissimopoulos <billziss at navimatics.com>.

    The full text of the GPLv3 license follows below.

I am hoping that this text addresses the issues brought up and that it allows fair use by true open-source projects.

**Comment by dumblob** [comment]
Sounds good @billziss-gh, thanks!

**Comment by billziss-gh** [comment]
Thank you everyone for the very helpful feedback. The license change has been checked in: 9c4b73dac19a8b42c81a7d2fa524e49a8045eb27

**Comment by slowmojoe1** [comment]
How does this license work when utilizing a project that utilizes WinFsp? For example, [rclone](https://rclone.org) is MIT licensed but it utilizes WinFsp which is GPLv3 licensed with this special clause. Are projects that utilize rclone now bound to GPLv3?

**Comment by megapro17** [comment]
It's not possible to create closed source file system for winfsp without buying commercial license, while for dokany you can do it?

**Comment by Masamune3210** [comment]
Because it's stipulated as such, creating a closed source project and using a open source project without compensation is dumb anyway

**Comment by megapro17** [comment]
It's indeed dumb that you can't use a tool for creating file systems and create your own filesystem without holding rights to your invention. Even linux fuse has lgpl license. This project is absolutely useless, even if i wanted to create gpl filesystem, i rather went with dokany. No surprise there's almost no fs's for winfsp.

**Comment by overheadhunter** [comment]
@megapro17 sorry but have you even considered asking for terms? WinFSP is an outstanding piece of software that offers substantial benefits for downstream projects and the conditions are very fair.

What are you trying to achieve here? What makes you so angry? Is it the fact that you're expected to pay for the work of others that your own commercial apps are based on?

It makes me sad how often people take exploiting open source software in commercial apps as a matter of course. When the upstream maintainers give up, dependent projects will die as well.

**Comment by Masamune3210** [comment]
You can make anything you want with it, you only need a commercial license if you want to interact commercially. If you don't like that I suggest you take a hard think about your mentality when it comes to fair rates for the work of others
